### PR TITLE
Use eslint-plugin-simple-import-sort

### LINKED ISSUE #51617: Restore case-sensitive import sorting
In #51579, I switched our sorting to be case insensitive, as sensitive sorting appeared to cause auto-imports to break.

This appears to be caused by #51616 (at least). After that's fixed, we could revert things.

Also, #51615, if we wanted to use `organize imports`, however that would be a problem regardless, but less of one given auto-imports should be able to put things in the right place. (Though, I'd rather just let a formatter do this for us.)

### PR DESCRIPTION
**Pull request by jakebailey**
This is an alternative to #51916, keeping sensitive sorting but correctly handling cases like `HasDecorators` and `hasDecorators` (which don't predicably sort) as well as sorting declarations themselves.

The sorting algorithm this uses is very simple: https://github.com/lydell/eslint-plugin-simple-import-sort/#sorting

~Unfortunately, this still doesn't enable using "organize imports" as that really changes things. But, this one is "fixable" and fully defines the ordering of all aspects of the imports for us, so `--fix` or a quickfix will always do the right thing.~

Organize imports now works as of #52115, I have added the unstable config to the settings template to see.

Fixes #51617

### CODE REVIEW

**Comment by jakebailey** [comment]
Found what is probably a bug in the new organize settings. With:

```json
    "typescript.unstable": {
        "organizeImportsCollation": "unicode",
        "organizeImportsCaseFirst": "upper",
        "organizeImportsIgnoreCase": false,
        "organizeImportsNumericCollation": true,
    },
```

Organize says:

```ts
export * from "../transformers/es2015";
export * from "../transformers/es2016";
export * from "../transformers/es2017";
export * from "../transformers/es2018";
export * from "../transformers/es2019";
export * from "../transformers/es2020";
export * from "../transformers/es2021";
export * from "../transformers/es5";
```

But simple-import-format says:

```ts
export * from "../transformers/es5";
export * from "../transformers/es2015";
export * from "../transformers/es2016";
export * from "../transformers/es2017";
export * from "../transformers/es2018";
export * from "../transformers/es2019";
export * from "../transformers/es2020";
export * from "../transformers/es2021";
```

The organizer was supposed to apply to paths too, so, I'll make a test case and look.

**Review by sheetalkamat — APPROVED** [review]

**Comment by jakebailey** [comment]
The other PR I plan to send after (maybe during?) this one is one that enforces the style of having a vertical import list. Right now, we have some that are vertical that I fixed myself, but some that aren't because they're newer. In lieu of a formatter which does this, the plugin is sort of needed to complete the "never have conflicts again" thing.

**Comment by jakebailey** [comment]
Just found one difference; our organize call moves the export statements in the Herebyfile too:

```patch
diff --git a/Herebyfile.mjs b/Herebyfile.mjs
index 4171dea36f..56b6005c3d 100644
--- a/Herebyfile.mjs
+++ b/Herebyfile.mjs
@@ -351,6 +351,10 @@ const { main: tsc, watch: watchTsc } = entrypointBuildTask({
     mainDeps: [generateLibs],
 });
 export { tsc, watchTsc };
+export { services, watchServices };
+export { tsserver, watchTsserver };
+export { lssl, watchLssl };
+export { tests, watchTests };
 
 
 const { main: services, build: buildServices, watch: watchServices } = entrypointBuildTask({
@@ -364,7 +368,6 @@ const { main: services, build: buildServices, watch: watchServices } = entrypoin
     mainDeps: [generateLibs],
     bundlerOptions: { exportIsTsObject: true },
 });
-export { services, watchServices };
 
 export const dtsServices = task({
     name: "dts-services",
@@ -388,7 +391,6 @@ const { main: tsserver, watch: watchTsserver } = entrypointBuildTask({
     output: "./built/local/tsserver.js",
     mainDeps: [generateLibs],
 });
-export { tsserver, watchTsserver };
 
 
 export const min = task({
@@ -417,7 +419,6 @@ const { main: lssl, build: buildLssl, watch: watchLssl } = entrypointBuildTask({
     mainDeps: [generateLibs],
     bundlerOptions: { exportIsTsObject: true },
 });
-export { lssl, watchLssl };
 
 export const dtsLssl = task({
     name: "dts-lssl",
@@ -455,7 +456,6 @@ const { main: tests, watch: watchTests } = entrypointBuildTask({
         }
     },
 });
-export { tests, watchTests };
 
 
 export const runEslintRulesTests = task({
```

**Comment by jakebailey** [comment]
Maybe this is the export logic not respecting the change we made to grouping for imports? Will file a bug.

**Review by sandersn — APPROVED** [review]

**Comment by jakebailey** [comment]
I'm going to hold off on this for now pending https://github.com/lydell/eslint-plugin-simple-import-sort/issues/124; I need to check to see if we have anything that fits this issue in our own repo as it'll end up breaking organize I'm pretty sure?

**Comment by jakebailey** [comment]
I'm going to mark this back to a draft; it turns out there are some edge cases with the `type` keyword that turn out not to agree between the tools. Given:

```ts
import {
    Apple,
    Pineapple,
    type Banana,
    Mango,
} from "foo";
```

TypeScript (with the unstable options) says:

```ts
import {
    Apple,
    Mango,
    Pineapple,
    type Banana,
} from "foo";
```

`simple-import-format` says:

```ts
import {
    type Banana,
    Apple,
    Mango,
    Pineapple,
} from "foo";
```

And `dprint` says:

```ts
import {
    Apple,
    type Banana,
    Mango,
    Pineapple,
} from "foo";
```

My personal feeling is that the last one is the "correct" one, as it's the one where you can add or remove a `type` keyword and the ordering doesn't change, but I also don't know if our behavior is intended. See also https://github.com/lydell/eslint-plugin-simple-import-sort/issues/124.

**Comment by jakebailey** [comment]
On second thought, I'm not going to mark this as a draft; we don't have any type keywords at the moment so I'm pretty sure that this PR is still an improvement on the current state of things, but it's something we should look into.

File: src/compiler/_namespaces/ts.moduleSpecifiers.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable simple-import-sort/exports */
```
Comment (jerone) [inline src/compiler/_namespaces/ts.moduleSpecifiers.ts:1]: Maybe add an override in `.eslintrc.json` for all files part of `/_namespaces/` path, to disable this specific rule.

Comment (jakebailey) [reply src/compiler/_namespaces/ts.moduleSpecifiers.ts:1]: That's not a bad idea, but I don't see us ever adding any more of these files anymore (in fact, I have plans to delete most of them)

**Comment by jakebailey** [comment]
As of `eslint-plugin-simple-import-sort` v10, there are now only two behaviors, the one used by TypeScript's current sorter (type imports at the end), and the one used by dprint/simple-import-sort (sort as though the type keyword was not present). Moving towards convergence.

**Comment by jakebailey** [comment]
Per offline discussion, I'm going to merge this as-is; I am working on adding configurability for the `type` keyword (it's tricky) and since we don't use any in the codebase at the moment, we can work on adding that later.

**Pull request merged**